lib/base-compiler.js

```diff
@@ -784,8 +821,33 @@ export class BaseCompiler {
         return result;
     }
 
-    getCacheKey(source, options, backendOptions, filters, tools, libraries) {
-        return {compiler: this.compiler, source, options, backendOptions, filters, tools, libraries};
+    getCacheKey(source, options, backendOptions, filters, tools, libraries, files) {
+        if (files) {
+            return {compiler: this.compiler, source, options, backendOptions, filters, tools, libraries, files};
+        } else {
+            return {compiler: this.compiler, source, options, backendOptions, filters, tools, libraries};
+        }
+    }
+
+    getCmakeCacheKey(key, files) {
+        const cacheKey = Object.assign({}, key);
+        cacheKey.compiler = this.compiler;
+        cacheKey.files = files;
+        cacheKey.api = 'cmake';
+
+        if (cacheKey.filters && cacheKey.filters.execute) {
+            delete cacheKey.filters.execute;
+        }
+
+        if (cacheKey.executionParameters) {
```

Comment:
TIL you can `delete` a non-existing key without error!

```js
> a = {}
< {}
> delete a.moo
< true
```
which can simplify some of these clauses (and probalby some of my original code where I do this too!)